codegen/masm/src/codegen/scheduler.rs

```diff
@@ -0,0 +1,1404 @@
+            self.block_info
+                .treegraph
+                .is_scheduled_after(td, this_dependent)
+        });
+        let is_live_after = match arg_node.into() {
+            Node::Argument(ArgumentNode::Direct { .. }) => successors
+                .map(|jts| {
+                    jts.iter().any(|jt| {
+                        let defined_by = self.f.dfg.block_args(jt.destination).contains(&arg);
+                        let is_live_at = self
+                            .liveness
+                            .is_live_at(&arg, ProgramPoint::Block(jt.destination));
+                        is_live_at && !defined_by
+                    })
+                })
+                .unwrap_or_else(|| {
+                    self.liveness
+                        .is_live_after(&arg, ProgramPoint::Block(self.block_info.source))
+                }),
+            Node::Argument(ArgumentNode::Indirect { successor, .. })
+            | Node::Argument(ArgumentNode::Conditional { successor, .. }) => {
+                let successors = successors.unwrap();
+                let successor = successors[successor as usize].destination;
+                let defined_by = self.f.dfg.block_args(successor).contains(&arg);
+                let is_live_at = self
+                    .liveness
+                    .is_live_at(&arg, ProgramPoint::Block(successor));
+                is_live_at && !defined_by
+            }
+            _ => unreachable!(),
+        };
+        let is_last_use = !is_live_after && is_last_dependent;
+        if is_last_use {
+            Constraint::Move
+        } else {
+            Constraint::Copy
+        }
+    }
+}
+
+fn transitive_instruction_dependents(
+    stack_or_result_node: NodeId,
+    current_block_info: &BlockInfo,
+) -> SmallSet<NodeId, 2> {
+    current_block_info
+        .depgraph
+        .predecessors(stack_or_result_node)
+        .map(|p| {
+            assert!(p.dependent.is_argument());
+            current_block_info.depgraph.unwrap_parent(p.dependent)
+        })
+        .collect()
+}
+
+fn build_dependency_graph(
+    block_id: hir::Block,
+    function: &hir::Function,
+    liveness: &LivenessAnalysis,
+) -> DependencyGraph {
+    let mut graph = DependencyGraph::default();
+
+    // This set represents the values which are guaranteed to be materialized for an instruction
+    let mut materialized_args = SmallSet::<hir::Value, 4>::default();
+    // This map represents values used as block arguments, and the successors which use them
+    let mut block_arg_uses = SmallMap::<hir::Value, SmallSet<hir::Block, 2>>::default();
+
+    // For each instruction, record it and it's arguments/results in the graph
+    for (inst_index, inst) in function.dfg.block_insts(block_id).enumerate() {
+        materialized_args.clear();
+        block_arg_uses.clear();
+
+        let node_id = graph.add_node(Node::Inst {
+            id: inst,
+            pos: inst_index as u16 + 1,
```

Comment:
I believe at least part of the reason I did so was to keep it in sync with the "real" position value of an instruction as computed by `DataFlowGraph::pp_cmp`, as at one point that was relevant - but I don't think that is actually necessary anymore after many of the other changes made. It can be zero-indexed without affecting anything now, at least AFAICT.